File: leetcode/js/sum-of-all-subset-xor-totals.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var subsetXORSum = function (nums) {
  const calcSubsetXor = (idx, sum) => {
    if (idx === nums.length) {
      return sum;
    }

    const addXor = calcSubsetXor(idx + 1, sum ^ nums[idx]);
    const noAddXor = calcSubsetXor(idx + 1, sum);

    return addXor + noAddXor;
  };

  return calcSubsetXor(0, 0);
};

var subsetXORSum = function (nums) {
  let result = 0;
  const calcSubsetXor = (idx, sum) => {
    result += sum;

    if (idx === nums.length) {
      return;
    }

    for (let i = idx; i < nums.length; i++) {
      calcSubsetXor(i + 1, sum ^ nums[i]);
    }
  };

  calcSubsetXor(0, 0);

  return result;
};

File: leetcode/python/sum-of-all-subset-xor-totals.py
class Solution:
    def subsetXORSum(self, nums: list[int]) -> int:
        result = []

        def add_XOR(idx, res) -> None:
            result.append(res)

            for i in range(idx, len(nums)):
                add_XOR(i + 1, res ^ nums[i])

        add_XOR(0, 0)
        
        return sum(result)